msg0 = "\n######################################################################\n"
msg1 = "Input forward read is missing. Execution stop.\n"
msg2 = "Forward read type is missing. Please select between fasta or fasta. \n"
msg3 = "Reverse read type is missing. Please select between fasta or fasta. \n"
msg4 = "Dataset Type is missing. Please select between genomic, control or selection. \n"
msg5 = "3\' primer of forward read is missing.Execution stop.\n"
msg6 = "5\' primer of forward read is missing.Execution stop.\n"
msg7 = "3\' primer of reverse read is missing.Execution stop.\n"
msg8 = "5\' primer of reverse read is missing.Execution stop.\n"
msg9 = "Output folder path is missing.Execution stop.\n"
msg10 = "Output ID path is missing.Execution stop.\n"
msg11 = "Fasta sequence is missing.Execution stop.\n"
msg12 = "Annotation File is missing.Execution stop.\n"
msg13 = "No permission to write in folder. Exit\n"
msg14 = "Setting parameters\n"
msg15 = "Sequencing input:\t"
msg16 = "\nInput file forward"
msg17 = "\nInput dataset forward:\t"
msg18 = "\nInput dataset reverse:\t"
msg19 = "\nInput dataset forward type:\t"
msg20 = "\nInput dataset reverse type:\t"
msg21 = "\nPrimer 5\' forward read:\t"
msg22 = "\nPrimer 3\' forward read:\t"
msg23 = "\nPrimer 5\' reverse read:\t"
msg24 = "\nPrimer 3\' reverse read:\t"
msg25 = "\nSample Name:\t"
msg26 = "\nSequence file:\t"
msg27 = "\nAnnotation file:\t"
msg28 = "\nInput dataset:\t"
msg29 = "\nInput dataset type:\t"
msg30 = "\nPrimer 5\' read:\t"
msg31 = "\nPrimer 3\' read:\t"
msg32 = "Error during checking Cutadapt installation. Execution stop.\n"
msg33 = "\nCutadapt version:\t"
msg34 = "Error during trimming 5\'primers. Exit\n"
msg35 = "\nTrimming of 5\' primers is complete."
msg36 = "Error during trimming 3\'primers. Exit\n"
msg37 = "Trimming of 3\' primers is complete.\n"
msg38a = "Error during trimming 5\'primers. Exit\n"
msg38b = "\nTrimming of 5\' primers of forward and reverse dataset is complete.\n"
msg39 = "Error during trimming 3\'primers. Exit\n"
msg40 = "\nTrimming of 3\' primers of forward and reverse dataset is complete.\n"
msg41 = "Error during annotation parsing.Execution stop.\n"
msg42 = "Annotation file parsing complete.\n"
msg43 = "Error during genome sequence parsing.Execution stop.\n"
msg44 = "Fasta sequence file parsing complete.\n"
msg45 = "Error during genome format file parsing.Execution stop.\n"
msg46 = "Genome format file parsing complete.\n"
msg47 = "\nForward read number:\t"
msg48 = "\nReverse read number:\t"
msg49 = "\nNumber of reads:\t"
msg50 = "Number of reads after 5\' trimming\t"
msg51 = "Number of reads after 3\' trimming\t"
msg52 = "Number of forward reads after 5\' trimming.\t"
msg53 = "\nNumber of reverse reads after 5\' trimming.\t"
msg54 = "Number of forward reads after 3\' trimming.\t"
msg55 = "\nNumber of reverse reads after 3\' trimming.\t"
msg56 = "Total reads ready for mapping.\t"
msg57 = "Database already exist. Exit\n"
msg58 = "Error makeblastdb. Exit\n"
msg59 = "Make blast db complete.\n"
msg60 = "Error during blastn analysis. Exit\n"
msg61 = "Blastn analysis complete."
msg62 = "Number of reads mapping:\t"
msg63 = "\nNumber of reads after open gap filtering:\t"
msg64 = "\nNumber of reads after mismatch filtering:\t"
msg65 = "\nNumber of reads after lenght filtering:\t"
msg66 = "\nNumber of reads after quality filtering:\t"
msg67 = "Error during checking Pick Otus. Execution stop\n"
msg68 = "Error during checking Pick Rep Seq. Execution stop\n"
msg69 = "Error during checking Samtools installation. Execution stop\n"
msg70 = "Error during checking AWK installation. Execution stop\n"
msg71 = "\nError clustering reads. Exit\n"
msg72 = "\nError during picking representative clones. Exit"
msg73 = "\nRepresentative clones picking complete."
msg74 = "\nRepresentative clones parsing complete."
msg75 = "\nChecking clones gene intersection failed. Execution stop."
msg76 = "\nChecking clones gene intersection complete."
msg77 = "\nFiltering Clones outside CDS."
msg78 = "\nError clones annotation.Execution stop."
msg79 = "\nClones annotation complete."
msg80 = "Error during blastp analysis. Exit\n"
msg81 = "\nBlastp analysis complete."
msg82 = "\nError during translation in six dataframe.Execution stop"
msg83 = "\nTranslation complete."
msg84 = "\nChromosome name not given. Execution stop."
msg85 = "\nNo input file given. Intesection error. Execution stop."
msg86 = "\nOnly one input file found. No intersection will be performed."

