msg0 = "######################################################################\n"
msg1 = "#Error#Input forward read is missing. Execution stop.\n"
msg2 = "#Error#Forward read type is missing. Please select between fasta or fasta. \n"
msg3 = "#Error#Reverse read type is missing. Please select between fasta or fasta. \n"
msg4 = "\nPrimers insert are not coupled.Execution stop."
msg5 = "#Error#3\' primer of forward read is missing.Execution stop.\n"
msg6 = "#Error#5\' primer of forward read is missing.Execution stop.\n"
msg7 = "#Error#3\' primer of reverse read is missing.Execution stop.\n"
msg8 = "#Error#5\' primer of reverse read is missing.Execution stop.\n"
msg9 = "#Error#Output folder path is missing.Execution stop.\n"
msg10 = "#Error#Output ID path is missing.Execution stop.\n"
msg11 = "#Error#Fasta sequence is missing.Execution stop.\n"
msg12 = "#Error#Annotation File is missing.Execution stop.\n"
msg13 = "#Error#No permission to write in folder. Exit\n"
msg14 = "Setting parameters\n"
msg15 = "Sequencing input:\t"
msg16 = "\nInput file forward"
msg17 = "\nInput dataset forward:\t"
msg18 = "\nInput dataset reverse:\t"
msg19 = "\nInput dataset forward type:\t"
msg20 = "\nInput dataset reverse type:\t"
msg21 = "\nPrimer 5\' forward read:\t"
msg22 = "\nPrimer 3\' forward read:\t"
msg23 = "\nPrimer 5\' reverse read:\t"
msg24 = "\nPrimer 3\' reverse read:\t"
msg25 = "\nSample Name:\t"
msg26 = "\nSequence file:\t"
msg27 = "\nFasta name file:\t"
msg28 = "\nInput dataset:\t"
msg29 = "\nInput dataset type:\t"
msg30 = "\nPrimer 5\' read:\t"
msg31 = "\nPrimer 3\' read:\t"
msg32 = "#Error#Error during checking Cutadapt installation. Execution stop.\n"
msg33 = "\nCutadapt version:\t"
msg34 = "#Error#Error during trimming 5\'primers. Exit\n"
msg35 = "Trimming of 5\' primers is complete.\n"
msg36 = "#Error#Error during trimming 3\'primers. Exit\n"
msg37 = "\nTrimming of 3\' primers is complete."
msg38a = "#Error#Error during trimming 5\'primers. Exit\n"
msg38b = "\nTrimming of 5\' primers of forward and reverse dataset is complete.\n"
msg39 = "#Error#Error during trimming 3\'primers. Exit\n"
msg40 = "\nTrimming of 3\' primers of forward and reverse dataset is complete.\n"
msg41 = "#Error#Error during annotation parsing.Execution stop.\n"
msg42 = "Annotation file parsing complete.\n"
msg43 = "\nError during bowtie2 indexing.Execution stop."
msg44 = "\nBowtie2 indexing complete."
msg45 = "#Error#Error during genome format file parsing.Execution stop.\n"
msg46 = "Genome format file parsing complete.\n"
msg47 = "\nForward read number:\t"
msg48 = "\nReverse read number:\t"
msg49 = "\nNumber of reads:\t"
msg50 = "\nNumber of reads after 5\' trimming\t"
msg51 = "Number of reads after 3\' trimming\t"
msg52 = "\nNumber of forward reads after 5\' trimming.\t"
msg53 = "\nNumber of reverse reads after 5\' trimming.\t"
msg54 = "Number of forward reads after 3\' trimming.\t"
msg55 = "\nNumber of reverse reads after 3\' trimming.\t"
msg56 = "\nTotal reads ready for mapping.\t"
msg57 = "Database already exist. Exit\n"
msg58 = "#Error#Error SAM header reading.Execution stop\n"
msg59 = "\nSAM header complete."
msg60 = "#Error#'Error during mapping Kallisto.Execution stop\n"
msg61 = "\nKallisto mapping complete."
msg62 = "Number of reads mapping:\t"
msg63 = "Error SAM parsing.Exit\n"
msg64 = "\nSAM parsing complete."
msg65 = "\nNumber of reads after lenght filtering:\t"
msg66 = "\nNumber of reads after quality filtering:\t"
msg67 = "#Error#Error during checking Pick Otus. Execution stop\n"
msg68 = "#Error#Error during checking Pick Rep Seq. Execution stop\n"
msg69 = "#Error#Error during checking Samtools installation. Execution stop\n"
msg70 = "#Error#Error during checking AWK installation. Execution stop\n"
msg71 = "\n#Error#Error clustering reads.Exit\n"
msg72 = "\n#Error#Error during picking representative clones.Exit"
msg73 = "\nDomain coordinates extraction complete."
msg74 = "\nDomain coordinates extraction failed. Execution stop."
msg75 = "\n#Error#Checking clones gene intersection failed. Execution stop."
msg76 = "\nChecking clones gene intersection complete."
msg77 = "\nFiltering Clones outside CDS."
msg78 = "\n#Error#Error clones annotation.Execution stop."
msg79 = "\nClones annotation complete."
msg80 = "#Error#Error Mismatch filtering failed.Execution stop\n"
msg81 = "\nMismatch filtering complete."
msg82 = "#Error#Error during conversion SAM to BAM.Execution stop\n"
msg83 = "\nConversion SAM to BAM is complete."
msg84 = "#Error#Error during sorting BAM file. Execution stop\n"
msg85 = "\nSorting of BAM file complete."
msg86 = "\nOnly one input file found. No intersection will be performed."
msg87 = "\n#Error#Gene Name is missing.Execution stop."
msg88 = '\n#Error#Error during merging files.Execution stop.'
msg89 = '\n#Error#Error in pybedtools. Error in computing depth-of-coverage.Execution stop.'
msg90 = '\nComputing the depth-of-coverage complete.'
msg91 = '\n#Error#Error in computing breadth-of-coverage.Execution stop.'
msg92 = '\nComputing the breadth-of-coverage complete.'
msg93 = '\nBam2bed complete.'
msg94 = '\n#Error#Error in pybedtools. Function bedtools bam2bed failed.Execution stop.'
msg95 = '\nRead group by Id complete.'
msg96 = "\n#Error#Read group by ID failed.Execution stop."
msg97 = "\nRead count complete."
msg98 = "\n#Error#Read count failed.Execution stop."
msg99 = "\n#Error#Error computing Max depth coverage.Execution stop."
msg100 = "\nMax depth coverage computing complete."
msg101 = "\n#Error#Percentile depth filtering failed.Execution stop."
msg102 = "\nPercentile depth filtering complete."
msg103 = "\n#Error#Raw definition of domains failed.Execution stop."
msg104 = "\nRaw definition of domains complete."
msg105 = "\n#Error#Error computing domain start and end.Execution stop."
msg106 = "\nComputing domain start and end complete."
msg107 = "\n#Error#Parsing output failed.Execution stop."
msg108 = "\nParsing output complete"
msg109 = "\n#Error#Error parsing input file.Execution stop."
msg110 = "\nParsing input file complete."
msg111 = "\n#Error#Bedtools intersect of common domains failed.Execution stop."
msg112 = "\nBedtools intersect of common domains complete."
msg113 = "\n#Error#Bedtools intersect of unique domains failed.Execution stop."
msg114 = "\nBedtools intersect of unique domains complete."
msg115 = "\n#Error#Parsing files for edgeR analysis failed.Execution stop."
msg116 = "\nParsing files for edgeR analysis complete."
msg117 = "\n#Error#Differential expression analysis failed.Execution stop."
msg118 = "\nDifferential expression analysis complete."
msg119 = "\n#Error#Error parsing output file with common domains.Execution stop."
msg120 = "\nParsing output file with common domains complete."
msg121 = "\n#Error#Error parsing output file with unique domains.Execution stop."
msg122 = "\nParsing output file with unique domains complete."
msg123 = "\nReference sequence index already exist."
msg124 = "\nParsing genome complete."
msg125 = "\nError#Conversion BED to BAM failed.Execution stop."
msg126 = "\nConversion BED to BAM complete."
msg127 = "\nSingle-end analysis start."
msg128 = " \nSingle-end analysis failed. Input parameters not correct.Execution stop."
msg129 = "\nPaired-end analysis start."
msg130 = "\nTrimming complete.Starting mapping."
msg131 = "\nEstimating average and standard deviation of fragment length failed.Execution stop."
msg132 = "\nEstimating average and standard deviation of fragment length complete"

