msg0 = "######################################################################\n"
msg1 = "#Error#Input forward read is missing. Execution stop.\n"
msg2 = "#Error#Forward read type is missing. Please select between fasta or fasta. \n"
msg3 = "#Error#Reverse read type is missing. Please select between fasta or fasta. \n"
msg4 = "#Error#Dataset Type is missing. Please select between genomic, control or selection. \n"
msg5 = "#Error#3\' primer of forward read is missing.Execution stop.\n"
msg6 = "#Error#5\' primer of forward read is missing.Execution stop.\n"
msg7 = "#Error#3\' primer of reverse read is missing.Execution stop.\n"
msg8 = "#Error#5\' primer of reverse read is missing.Execution stop.\n"
msg9 = "#Error#Output folder path is missing.Execution stop.\n"
msg10 = "#Error#Output ID path is missing.Execution stop.\n"
msg11 = "#Error#Fasta sequence is missing.Execution stop.\n"
msg12 = "#Error#Annotation File is missing.Execution stop.\n"
msg13 = "#Error#No permission to write in folder. Exit\n"
msg14 = "Setting parameters\n"
msg15 = "Sequencing input:\t"
msg16 = "\nInput file forward"
msg17 = "\nInput dataset forward:\t"
msg18 = "\nInput dataset reverse:\t"
msg19 = "\nInput dataset forward type:\t"
msg20 = "\nInput dataset reverse type:\t"
msg21 = "\nPrimer 5\' forward read:\t"
msg22 = "\nPrimer 3\' forward read:\t"
msg23 = "\nPrimer 5\' reverse read:\t"
msg24 = "\nPrimer 3\' reverse read:\t"
msg25 = "\nSample Name:\t"
msg26 = "\nSequence file:\t"
msg27 = "\nFasta name file:\t"
msg28 = "\nInput dataset:\t"
msg29 = "\nInput dataset type:\t"
msg30 = "\nPrimer 5\' read:\t"
msg31 = "\nPrimer 3\' read:\t"
msg32 = "#Error#Error during checking Cutadapt installation. Execution stop.\n"
msg33 = "\nCutadapt version:\t"
msg34 = "#Error#Error during trimming 5\'primers. Exit\n"
msg35 = "\nTrimming of 5\' primers is complete."
msg36 = "#Error#Error during trimming 3\'primers. Exit\n"
msg37 = "Trimming of 3\' primers is complete.\n"
msg38a = "#Error#Error during trimming 5\'primers. Exit\n"
msg38b = "\nTrimming of 5\' primers of forward and reverse dataset is complete.\n"
msg39 = "#Error#Error during trimming 3\'primers. Exit\n"
msg40 = "\nTrimming of 3\' primers of forward and reverse dataset is complete.\n"
msg41 = "#Error#Error during annotation parsing.Execution stop.\n"
msg42 = "Annotation file parsing complete.\n"
msg43 = "#Error#Error during genome sequence parsing.Execution stop.\n"
msg44 = "Fasta sequence file parsing complete.\n"
msg45 = "#Error#Error during genome format file parsing.Execution stop.\n"
msg46 = "Genome format file parsing complete.\n"
msg47 = "\nForward read number:\t"
msg48 = "\nReverse read number:\t"
msg49 = "\nNumber of reads:\t"
msg50 = "\nNumber of reads after 5\' trimming\t"
msg51 = "Number of reads after 3\' trimming\t"
msg52 = "\nNumber of forward reads after 5\' trimming.\t"
msg53 = "\nNumber of reverse reads after 5\' trimming.\t"
msg54 = "Number of forward reads after 3\' trimming.\t"
msg55 = "\nNumber of reverse reads after 3\' trimming.\t"
msg56 = "Total reads ready for mapping.\t"
msg57 = "Database already exist. Exit\n"
msg58 = "#Error#Error SAM header reading.Execution stop\n"
msg59 = "SAM header complete.\n"
msg60 = "#Error#'Error during mapping Bowtie2.Execution stop\n"
msg61 = "Bowtie2 mapping complete.\n"
msg62 = "Number of reads mapping:\t"
msg63 = "Error SAM parsing.Exit\n"
msg64 = "SAM parsing complete.\n"
msg65 = "\nNumber of reads after lenght filtering:\t"
msg66 = "\nNumber of reads after quality filtering:\t"
msg67 = "#Error#Error during checking Pick Otus. Execution stop\n"
msg68 = "#Error#Error during checking Pick Rep Seq. Execution stop\n"
msg69 = "#Error#Error during checking Samtools installation. Execution stop\n"
msg70 = "#Error#Error during checking AWK installation. Execution stop\n"
msg71 = "\n#Error#Error clustering reads.Exit\n"
msg72 = "\n#Error#Error during picking representative clones.Exit"
msg73 = "\nRepresentative clones picking complete."
msg74 = "\nRepresentative clones parsing complete."
msg75 = "\n#Error#Checking clones gene intersection failed. Execution stop."
msg76 = "\nChecking clones gene intersection complete."
msg77 = "\nFiltering Clones outside CDS."
msg78 = "\n#Error#Error clones annotation.Execution stop."
msg79 = "\nClones annotation complete."
msg80 = "#Error#Error Mismatch filtering failed.Execution stop\n"
msg81 = "Mismatch filtering complete.\n"
msg82 = "#Error#Error during conversion SAM to BAM.Execution stop\n"
msg83 = "Conversion SAM to BAM is complete.\n"
msg84 = "#Error#Error during sorting BAM file. Execution stop\n"
msg85 = "Sorting of BAM file complete.\n"
msg86 = "\nOnly one input file found. No intersection will be performed."
msg87 = "\n#Error#Gene Name is missing.Execution stop."
msg88 = '\n#Error#Error during merging files.Execution stop.'
msg89 = '\n#Error#Error in pybedtools. Function bedtools getfasta failed.Execution stop.'
msg90 = '\n#Error#Error in pybedtools. Function bedtools getfasta failed.Execution stop.'
msg91 = '\nPybedtools getfasta complete.'
msg92 = '\n#Error#Error while cleaning balstn output.Execution stop.'
msg93 = '\nBlastn output hash clean complete.\n'
msg94 = '#Error#Error conversion FastQ ==>FastA.Execution stop.\n'
msg95 = 'Conversion FastQ ==>FastA complete.\n'
msg96 = "#Error#Error conversion FastA ==>Tabular.Execution stop.\n"
msg97 = "Conversion FastA ==>Tabular complete.\n"
msg98 = "#Error#Error renaming Fasta Id.Execution stop.\n"
msg99 = "Renaming Fasta Id complete.\n"
msg100 = "#Error#Error conversion Tabular ==>FastA.Execution stop.\n"
msg101 = "Conversion Tabular ==>FastA complete.\n"
msg102 = "\n#Error# Blastn Output filtering failed.Execution stop."
msg103 = "\nBlastn Output filtering complete."
msg104 = "\n#Error#Output Bed parsed failed.Execution stop."
msg105 = "\nOutput Bed parsed complete."
msg106 = "\n#Error#Error during clone cluster count.Execution stop."
msg107 = "\nClone cluster count.Complete"
msg108 = "\n#Error#Filtering cluster failed.Execution stop."
msg109 = "\nFiltering cluster complete."
msg110 = "\n#Error# Add gene description failed.Execution stop."
msg111 = "\nAdd gene description complete."
msg112 = "\n#Error#Add nucleotide sequence failed.Execution stop."
msg113 = "\nAdd nucleotide sequence complete."
msg114 = "Background input file:\t"
msg115 = "\nTarget input file:\t"
msg116 = "\nOutput ID:\t"
msg117 = "\nFasta input file:\t"
msg118 = "\n#Error#Conversion TAB ==> BED failed.Execution stop."
msg119 = "\nConversion TAB ==> BED complete."
msg120 = "\n#Error#Gene name missing.Execution stop."
msg121 = "\n#Error#Count gene lenght failed.Execution stop."
msg122 = "\nCount gene lenght complete."
msg123 = "\n#Error#Error parsing genome file.Execution stop."
msg124 = "\nParsing genome complete."
msg125 = "\nError#Conversion BED to BAM failed.Execution stop."
msg126 = "\nConversion BED to BAM complete."
msg127 = "\nError#Computing average depth BAM file failed.Execution stop."
msg128 = "\nComputing average depth BAM file complete."
msg129 = "\nError#Downsampling failed.Execution stop."
msg130 = "\nDownsampling complete."
msg131 = "\nError#Computiong coverage failed.Execution stop."
msg132 = "\nComputiong coverage complete."
msg133 = "\nError#Get intervals with positive coverage failed.Execution stop."
msg134 = "\nGet intervals with positive coverage complete."
msg135 = "\nError#Parsing domains failed.Execution stop."
msg136 = "\nParsing domains complete."
msg137 = "\nError#Intervals parsing failed.Execution stop."
msg138 = "\nIntervals parsing coverage complete."