msg0 = "\n###-------------------------------------------------------------------\n"
msg1 = "\n@@@Input forward read is missing. Execution stop.\n"
msg2 = "\n@@@Forward read type is missing. Please select between fasta or fasta.\n"
msg3 = "\n@@@Reverse read type is missing. Please select between fasta or fasta.\n"
msg4 = "\n@@@Primers insert are not coupled. Execution stop.\n"
msg5 = "\n@@@3\' primer of forward read is missing. Execution stop.\n"
msg6 = "\n@@@5\' primer of forward read is missing. Execution stop.\n"
msg7 = "\n@@@3\' primer of reverse read is missing. Execution stop.\n"
msg8 = "\n@@@5\' primer of reverse read is missing. Execution stop.\n"
msg9 = "\n@@@Output folder path is missing. Execution stop.\n"
msg10 = "\n@@@Output ID path is missing. Execution stop.\n"
msg11 = "\n@@@Fasta sequence is missing. Execution stop.\n"
msg12 = "\n@@@Annotation File is missing. Execution stop.\n"
msg13 = "\n@@@No permission to write in folder. Execution stop.\n"
msg14 = "\n###Setting Parameters\n"
msg15 = "\n###Sequencing input:\t"
msg16 = "\n###Input file forward\n"
msg17 = "\n###Input dataset forward:\t"
msg18 = "\n###Input dataset reverse:\t"
msg19 = "\n###Input dataset forward type:\t"
msg20 = "\n###Input dataset reverse type:\t"
msg21 = "\n###Primer 5\' forward read:\t"
msg22 = "\n###Primer 3\' forward read:\t"
msg23 = "\n###Primer 5\' reverse read:\t"
msg24 = "\n###Primer 3\' reverse read:\t"
msg25 = "\n###Sample Name:\t"
msg26 = "\n###Sequence file:\t"
msg27 = "\n###Fasta name file:\t"
msg28 = "\n###Input dataset:\t"
msg29 = "\n###Input dataset type:\t"
msg30 = "\n###Primer 5\' read:\t"
msg31 = "\n###Primer 3\' read:\t"
msg32 = "\n@@@Error during checking Cutadapt installation. Execution stop.\n"
msg33 = "\n###Cutadapt version:\t"
msg34 = "\n@@@Error during trimming 5\'primers. Execution stop.\n"
msg35 = "\n###Trimming of 5\' primers is complete.\n"
msg36 = "\n@@@Error during trimming 3\'primers. Execution stop.\n"
msg37 = "\n###Trimming of 3\' primers is complete.\n"
msg38a = "\n@@@Error during trimming 5\'primers. Execution stop.\n"
msg38b = "\n###Trimming of 5\' primers of forward and reverse dataset is complete.\n"
msg39 = "\n@@@Error during trimming 3\'primers. Execution stop.\n"
msg40 = "\n###Trimming of 3\' primers of forward and reverse dataset is complete.\n"
msg41 = "\n@@@Error Trimming. Execution stop.\n"
msg43 = "\n@@@Error during bowtie2 indexing. Execution stop.\n"
msg44 = "\n###Bowtie2 indexing complete.\n"
msg45 = "\n@@@Error during genome format file parsing. Execution stop.\n"
msg46 = "\n###Genome format file parsing complete.\n"
msg47 = "\n###Forward read number:\t"
msg48 = "\n###Reverse read number:\t"
msg49 = "\n###Number of reads:\t"
msg50 = "\n###Number of reads after 5\' trimming\t"
msg51 = "\n###Number of reads after 3\' trimming\t"
msg52 = "\n###Number of forward reads after 5\' trimming.\t"
msg53 = "\n###Number of reverse reads after 5\' trimming.\t"
msg54 = "\n###Number of forward reads after 3\' trimming.\t"
msg55 = "\n###Number of reverse reads after 3\' trimming.\t"
msg56 = "\n###Total reads ready for mapping.\t"
msg57 = "\n###Database already exist. Execution stop.\n"
msg58 = "\n@@@Error SAM header reading. Execution stop.\n"
msg59 = "\n###SAM header complete.\n"
msg60 = "\n@@@Error during mapping Kallisto. Execution stop.\n"
msg61 = "\n###Kallisto mapping complete.\n"
msg62 = "\n###Number of reads mapping:\t"
msg63 = "\n@@@Error SAM parsing. Execution stop.\n"
msg64 = "\n###SAM parsing complete.\n"
msg65 = "\n###Number of reads after lenght filtering:\t"
msg66 = "\n###Number of reads after quality filtering:\t"
msg67 = "\n@@@Error during checking Pick Otus. Execution stop.\n"
msg68 = "\n@@@Error during checking Pick Rep Seq. Execution stop.\n"
msg69 = "\n@@@Error during checking Samtools installation. Execution stop.\n"
msg70 = "\n@@@Error during checking AWK installation. Execution stop.\n"
msg71 = "\n@@@Error clustering reads. Execution stop.\n"
msg72 = "\n@@@Error during picking representative clones. Execution stop.\n"
msg73 = "\n###Domain coordinates extraction complete.\n"
msg74 = "\n###Domain coordinates extraction failed. Execution stop.\n"
msg75 = "\n@@@Checking clones gene intersection failed. Execution stop.\n"
msg76 = "\n###Checking clones gene intersection complete.\n"
msg77 = "\n###Filtering Clones outside CDS.\n"
msg78 = "\n@@@Error clones annotation. Execution stop.\n"
msg79 = "\n###Clones annotation complete.\n"
msg80 = "\n@@@Error Mismatch filtering failed. Execution stop.\n"
msg81 = "\n###Mismatch filtering complete.\n"
msg82 = "\n@@@Error during conversion SAM to BAM. Execution stop.\n"
msg83 = "\n###Conversion SAM to BAM is complete.\n"
msg84 = "\n@@@Error during sorting BAM file. Execution stop.\n"
msg85 = "\n###Sorting of BAM file complete.\n"
msg86 = "\n###Only one input file found. No intersection will be performed.\n"
msg87 = "\n@@@Gene Name is missing. Execution stop.\n"
msg88 = "\n@@@Error during merging files. Execution stop.\n"
msg89 = "\n@@@Error in pybedtools. Error in computing depth-of-coverage. Execution stop.\n"
msg90 = "\n###Computing the depth-of-coverage complete.\n"
msg91 = "\n@@@Error in computing breadth-of-coverage. Execution stop.\n"
msg92 = "\n###Computing the breadth-of-coverage complete.\n"
msg93 = "\n###Bam2bed complete.\n"
msg94 = "\n@@@Error in pybedtools. Function bedtools bam2bed failed. Execution stop.\n"
msg95 = "\n###Read group by Id complete.\n"
msg96 = "\n@@@Read group by ID failed. Execution stop.\n"
msg97 = "\n###Read count complete.\n"
msg98 = "\n@@@Read count failed. Execution stop.\n"
msg99 = "\n@@@Error computing Max depth coverage. Execution stop.\n"
msg100 = "\n###Max depth coverage computing complete.\n"
msg101 = "\n@@@Percentile depth filtering failed. Execution stop.\n"
msg102 = "\n###Percentile depth filtering complete.\n"
msg103 = "\n@@@Raw definition of domains failed. Execution stop.\n"
msg104 = "\n###Raw definition of domains complete.\n"
msg105 = "\n@@@Error computing domain start and end. Execution stop.\n"
msg106 = "\n###Computing domain start and end complete.\n"
msg107 = "\n@@@Parsing output failed. Execution stop.\n"
msg108 = "\n###Parsing output complete"
msg109 = "\n@@@Error parsing input file. Execution stop.\n"
msg110 = "\n###Parsing input file complete.\n"
msg111 = "\n@@@Bedtools intersect of common domains failed. Execution stop.\n"
msg112 = "\n###Bedtools intersect of common domains complete.\n"
msg113 = "\n@@@Bedtools intersect of unique domains failed. Execution stop.\n"
msg114 = "\n###Bedtools intersect of unique domains complete.\n"
msg115 = "\n@@@Parsing files for edgeR analysis failed. Execution stop.\n"
msg116 = "\n###Parsing files for edgeR analysis complete.\n"
msg117 = "\n@@@Differential expression analysis failed. Execution stop.\n"
msg118 = "\n###Differential expression analysis complete.\n"
msg119 = "\n@@@Error parsing output file with common domains. Execution stop.\n"
msg120 = "\n###Parsing output file with common domains complete.\n"
msg121 = "\n@@@Error parsing output file with unique domains. Execution stop.\n"
msg122 = "\n###Parsing output file with unique domains complete.\n"
msg123 = "\n###Reference sequence index already exist.\n"
msg124 = "\n###Parsing genome complete.\n"
msg125 = "\n###Error#Conversion BED to BAM failed. Execution stop.\n"
msg126 = "\n###Conversion BED to BAM complete.\n"
msg127 = "\n###Single-end analysis start.\n"
msg128 = "\n###Single-end analysis failed. Input parameters not correct. Execution stop.\n"
msg129 = "\n###Paired-end analysis start.\n"
msg130 = "\n###Trimming complete. Starting mapping.\n"
msg131 = "\n###Estimating average and standard deviation of fragment length failed. Execution stop.\n"
msg132 = "\n###Estimating average and standard deviation of fragment length complete.\n"
