msg0 = "\n###-------------------------------------------------------------------\n"
msg1 = "\n@@@Input forward read is missing. Execution stop.\n"
msg2 = "\n@@@Forward read type is missing. Please select between fasta or fasta.\n"
msg3 = "\n@@@Reverse read type is missing. Please select between fasta or fasta.\n"
msg4 = "\n@@@Dataset Type is missing. Please select between genomic, control or selection.\n"
msg5 = "\n@@@3\' primer of forward read is missing. Execution stop.\n"
msg6 = "\n@@@5\' primer of forward read is missing. Execution stop.\n"
msg7 = "\n@@@3\' primer of reverse read is missing. Execution stop.\n"
msg8 = "\n@@@5\' primer of reverse read is missing. Execution stop.\n"
msg9 = "\n@@@Output folder path is missing. Execution stop.\n"
msg10 = "\n@@@Output ID path is missing. Execution stop.\n"
msg11 = "\n@@@Fasta sequence is missing. Execution stop.\n"
msg12 = "\n@@@Annotation File is missing. Execution stop.\n"
msg13 = "\n@@@No permission to write in folder. Execution stop.\n"
msg14 = "\n###Setting Parameters\n"
msg15 = "\n###Sequencing input:\t"
msg16 = "\n###Input file forward\n"
msg17 = "\n###Input dataset forward:\t"
msg18 = "\n###Input dataset reverse:\t"
msg19 = "\n###Input dataset forward type:\t"
msg20 = "\n###Input dataset reverse type:\t"
msg21 = "\n###Primer 5\' forward read:\t"
msg22 = "\n###Primer 3\' forward read:\t"
msg23 = "\n###Primer 5\' reverse read:\t"
msg24 = "\n###Primer 3\' reverse read:\t"
msg25 = "\n###Sample Name:\t"
msg26 = "\n###Sequence file:\t"
msg27 = "\n###Fasta name file:\t"
msg28 = "\n###Input dataset:\t"
msg29 = "\n###Input dataset type:\t"
msg30 = "\n###Primer 5\' read:\t"
msg31 = "\n###Primer 3\' read:\t"
msg32 = "\n@@@Error during checking Cutadapt installation. Execution stop.\n"
msg33 = "\n###Cutadapt version:\t"
msg34 = "\n@@@Error during trimming 5\'primers. Execution stop.\n"
msg35 = "\n###Trimming of 5\' primers is complete.\n"
msg36 = "\n@@@Error during trimming 3\'primers. Execution stop.\n"
msg37 = "\n###Trimming of 3\' primers is complete.\n"
msg38a = "\n@@@Error during trimming 5\'primers. Execution stop.\n"
msg38b = "\n###Trimming of 5\' primers of forward and reverse dataset is complete.\n"
msg39 = "\n@@@Error during trimming 3\'primers. Execution stop.\n"
msg40 = "\n###Trimming of 3\' primers of forward and reverse dataset is complete.\n"
msg41 = "\n@@@Error during annotation parsing. Execution stop.\n"
msg42 = "\n###Annotation file parsing complete.\n"
msg43 = "\n@@@Error during genome sequence parsing. Execution stop.\n"
msg44 = "\n###Fasta sequence file parsing complete.\n"
msg45 = "\n@@@Error during genome format file parsing. Execution stop.\n"
msg46 = "\n###Genome format file parsing complete.\n"
msg47 = "\n###Forward read number:\t"
msg48 = "\n###Reverse read number:\t"
msg49 = "\n###Number of reads:\t"
msg50 = "\n###Number of reads after 5\' trimming\t"
msg51 = "\n###Number of reads after 3\' trimming\t"
msg52 = "\n###Number of forward reads after 5\' trimming.\t"
msg53 = "\n###Number of reverse reads after 5\' trimming.\t"
msg54 = "\n###Number of forward reads after 3\' trimming.\t"
msg55 = "\n###Number of reverse reads after 3\' trimming.\t"
msg56 = "\n###Total reads ready for mapping.\t"
msg57 = "\n###Database already exist. Execution stop.\n"
msg58 = "\n###Error makeblastdb. Execution stop.\n"
msg59 = "\n###Make blast db complete.\n"
msg60 = "\n@@@Error during blastn analysis. Execution stop.\n"
msg61 = "\n###Blastn analysis complete.\n"
msg62 = "\n###Number of reads mapping:\t"
msg63 = "\n###Number of reads after open gap filtering:\t"
msg64 = "\n###Number of reads after mismatch filtering:\t"
msg65 = "\n###Number of reads after lenght filtering:\t"
msg66 = "\n###Number of reads after quality filtering:\t"
msg67 = "\n@@@Error during checking Pick Otus. Execution stop.\n"
msg68 = "\n@@@Error during checking Pick Rep Seq. Execution stop.\n"
msg69 = "\n@@@Error during checking Samtools installation. Execution stop.\n"
msg70 = "\n@@@Error during checking AWK installation. Execution stop.\n"
msg71 = "\n@@@Error clustering reads. Execution stop.\n"
msg72 = "\n@@@Error during picking representative clones. Execution stop.\n"
msg73 = "\n###Representative clones picking complete.\n"
msg74 = "\n###Representative clones parsing complete.\n"
msg75 = "\n@@@Checking clones gene intersection failed. Execution stop.\n"
msg76 = "\n###Checking clones gene intersection complete.\n"
msg77 = "\n###Filtering Clones outside CDS.\n"
msg78 = "\n@@@Error clones annotation. Execution stop.\n"
msg79 = "\n###Clones annotation complete.\n"
msg80 = "\n@@@Error during blastp analysis. Execution stop.\n"
msg81 = "\n###Blastp analysis complete.\n"
msg82 = "\n@@@Error during translation in six dataframe. Execution stop.\n"
msg83 = "\n###Translation complete.\n"
msg84 = "\n@@@Chromosome name not given. Execution stop.\n"
msg85 = "\n@@@No input file given. Intesection error. Execution stop.\n"
msg86 = "\n###Only one input file found. No intersection will be performed.\n"
msg87 = "\n@@@Gene Name is missing. Execution stop.\n"
msg88 = "\n@@@Error during merging files. Execution stop.\n"
msg89 = "\n@@@Error in pybedtools. Function bedtools getfasta failed. Execution stop.\n"
msg90 = "\n@@@Error in pybedtools. Function bedtools getfasta failed. Execution stop.\n"
msg91 = "\n###Pybedtools getfasta complete.\n"
msg92 = "\n@@@Error while cleaning balstn output. Execution stop.\n"
msg93 = "\n###Blastn output hash clean complete.\n"
msg94 = "\n@@@Error conversion FastQ ==> FastA. Execution stop.\n"
msg95 = "\n###Conversion FastQ ==> FastA complete.\n"
msg96 = "\n@@@Error conversion FastA ==> Tabular. Execution stop.\n"
msg97 = "\n###Conversion FastA ==> Tabular complete.\n"
msg98 = "\n@@@Error renaming Fasta Id. Execution stop.\n"
msg99 = "\n###Renaming Fasta Id complete.\n"
msg100 = "\n@@@Error conversion Tabular ==> FastA. Execution stop.\n"
msg101 = "\n###Conversion Tabular ==> FastA complete.\n"
msg102 = "\n@@@ Blastn Output filtering failed. Execution stop.\n"
msg103 = "\n###Blastn Output filtering complete.\n"
msg104 = "\n@@@Output Bed parsed failed. Execution stop.\n"
msg105 = "\n###Output Bed parsed complete.\n"
msg106 = "\n@@@Error during clone cluster count. Execution stop.\n"
msg107 = "\n###Clone cluster count.Complete"
msg108 = "\n@@@Filtering cluster failed. Execution stop.\n"
msg109 = "\n###Filtering cluster complete.\n"
msg110 = "\n@@@Add gene description failed. Execution stop.\n"
msg111 = "\n###Add gene description complete.\n"
msg112 = "\n@@@Add nucleotide sequence failed. Execution stop.\n"
msg113 = "\n###Add nucleotide sequence complete.\n"
msg114 = "\n###Background input file:\t"
msg115 = "\n###Target input file:\t"
msg116 = "\n###Output ID:\t"
msg117 = "\n###Fasta input file:\t"
msg118 = "\n@@@Conversion TAB ==> BED failed. Execution stop.\n"
msg119 = "\n###Conversion TAB ==> BED complete.\n"
msg120 = "\n@@@Gene name missing. Execution stop.\n"
msg121 = "\n@@@Count gene lenght failed. Execution stop.\n"
msg122 = "\n###Count gene lenght complete.\n"
msg123 = "\n@@@Error parsing genome file. Execution stop.\n"
msg124 = "\n###Parsing genome complete.\n"
msg125 = "\n###Error#Conversion BED to BAM failed. Execution stop.\n"
msg126 = "\n###Conversion BED to BAM complete.\n"
msg127 = "\n@@@Computing average depth BAM file failed. Execution stop.\n"
msg128 = "\n###Computing average depth BAM file complete.\n"
msg129 = "\n@@@Downsampling failed. Execution stop.\n"
msg130 = "\n###Downsampling complete.\n"
msg131 = "\n@@@Computiong coverage failed. Execution stop.\n"
msg132 = "\n###Computiong coverage complete.\n"
msg133 = "\n@@@Get intervals with positive coverage failed. Execution stop.\n"
msg134 = "\n###Get intervals with positive coverage complete.\n"
msg135 = "\n@@@Parsing domains failed. Execution stop.\n"
msg136 = "\n###Parsing domains complete.\n"
msg137 = "\n@@@Intervals parsing failed. Execution stop.\n"
msg138 = "\n###Intervals parsing coverage complete.\n"
msg139 = "\n@@@Parsing of mapping output file failed. Execution stop.\n"
msg140 = "\n###Parsing of mapping output file complete.\n"
msg141 = "\n@@@Parsing of domain definition output file failed. Execution stop.\n"
msg142 = "\n###Parsing of domain definition output file complete.\n"
msg143 = "\n@@@Bedtools coverage failed. Execution stop.\n"
msg144 = "\n###Bedtools coverage complete.\n"
msg145 = "\n@@@Bedtools coverage failed. Execution stop.\n"
msg146 = "\n###Bedtools coverage complete.\n"
msg147 = "\n@@@Parsing output bedtools coverage failed. Execution stop.\n"
msg148 = "\n###Parsing output bedtools coverage complete.\n"
msg149 = "\n@@@Differential expression analysis failed. Execution stop.\n"
msg150 = "\n###Differential expression analysis complete.\n"
msg151 = "\n@@@Parsing output edgeR failed. Execution stop.\n"
msg152 = "\n###Parsing output edgeR complete.\n"
msg153 = "\n@@@Substraction domains failed. Execution stop.\n"
msg154 = "\n###Substraction domains complete.\n"
msg155 = "\n@@@Intersection domains failed. Execution stop.\n"
msg156 = "\n###Intersection domains complete.\n"
