msg0 = "######################################################################\n"
msg1 = "#Error#Input forward read is missing. Execution stop.\n"
msg2 = "#Error#Forward read type is missing. Please select between fasta or fasta. \n"
msg3 = "#Error#Reverse read type is missing. Please select between fasta or fasta. \n"
msg4 = "\nAnnotation file:\t"
msg5 = "#Error#Forward dataset is in FastQ format while Reverse dataset is in FastA format.Execution stop.\n"
msg6 = "#Error#Forward dataset is in FastA format while Reverse dataset is in FastQ format.Execution stop.\n"
msg7 = "Reference parsing to tabular complete.\n"
msg8 = "#Error#Reference parsing to tabular failed.Execution stop.\n"
msg9 = "#Error#Output folder path is missing.Execution stop.\n"
msg10 = "#Error#Output ID path is missing.Execution stop.\n"
msg11 = "#Error#Fasta sequence is missing.Execution stop.\n"
msg12 = "#Error#Annotation File is missing.Execution stop.\n"
msg13 = "#Error#No permission to write in folder. Exit\n"
msg14 = "Setting parameters\n"
msg15 = "Sequencing input:\t"
msg16 = "\nInput file forward"
msg17 = "\nInput dataset forward:\t"
msg18 = "\nInput dataset reverse:\t"
msg19 = "\nInput dataset forward type:\t"
msg20 = "\nInput dataset reverse type:\t"
msg21 = "\nPrimer 5\' forward read:\t"
msg22 = "\nPrimer 3\' forward read:\t"
msg23 = "\nPrimer 5\' reverse read:\t"
msg24 = "\nPrimer 3\' reverse read:\t"
msg25 = "\nSample Name:\t"
msg26 = "\nSequence file:\t"
msg27 = "\nAnnotation file:\t"
msg28 = "\nInput dataset:\t"
msg29 = "\nInput dataset type:\t"
msg30 = "\nPrimer 5\' read:\t"
msg31 = "\nPrimer 3\' read:\t"
msg32 = "#Error#Error during checking Cutadapt installation. Execution stop.\n"
msg33 = "\nCutadapt version:\t"
msg34 = "#Error#Error during trimming 5\'primers. Exit\n"
msg35 = "\nTrimming of 5\' primers is complete."
msg36 = "#Error#Error during trimming 3\'primers. Exit\n"
msg37 = "Trimming of 3\' primers is complete.\n"
msg38a = "#Error#Error during trimming 5\'primers. Exit\n"
msg38b = "\nTrimming of 5\' primers of forward and reverse dataset is complete.\n"
msg39 = "#Error#Error during trimming 3\'primers. Exit\n"
msg40 = "\nTrimming of 3\' primers of forward and reverse dataset is complete.\n"
msg41 = "#Error#Error during annotation parsing.Execution stop.\n"
msg42 = "Annotation file parsing complete.\n"
msg43 = "#Error#Error during genome sequence parsing.Execution stop.\n"
msg44 = "Fasta sequence file parsing complete.\n"
msg45 = "#Error#Error during genome format file parsing.Execution stop.\n"
msg46 = "Genome format file parsing complete.\n"
msg47 = "\nForward read number:\t"
msg48 = "\nReverse read number:\t"
msg49 = "\nNumber of reads:\t"
msg50 = "Number of reads after 5\' trimming\t"
msg51 = "Number of reads after 3\' trimming\t"
msg52 = "Number of forward reads after 5\' trimming.\t"
msg53 = "\nNumber of reverse reads after 5\' trimming.\t"
msg54 = "Number of forward reads after 3\' trimming.\t"
msg55 = "\nNumber of reverse reads after 3\' trimming.\t"
msg56 = "Total reads ready for mapping.\t"
msg57 = "Database already exist. Exit\n"
msg58 = "Error makeblastdb. Exit\n"
msg59 = "Make blast db complete.\n"
msg60 = "#Error#Error during blastn analysis. Exit\n"
msg61 = "Blastn analysis complete."
msg62 = "Number of reads mapping:\t"
msg63 = "\nNumber of reads after open gap filtering:\t"
msg64 = "\nNumber of reads after mismatch filtering:\t"
msg65 = "\nNumber of reads after lenght filtering:\t"
msg66 = "\nNumber of reads after quality filtering:\t"
msg67 = "#Error#Error during checking Pick Otus. Execution stop\n"
msg68 = "#Error#Error during checking Pick Rep Seq. Execution stop\n"
msg69 = "#Error#Error during checking Samtools installation. Execution stop\n"
msg70 = "#Error#Error during checking AWK installation. Execution stop\n"
msg71 = "\n#Error#Error clustering reads.Exit\n"
msg72 = "\n#Error#Error during picking representative clones.Exit"
msg73 = "\nRepresentative clones picking complete."
msg74 = "\nRepresentative clones parsing complete."
msg75 = "\n#Error#Checking clones gene intersection failed. Execution stop."
msg76 = "\nChecking clones gene intersection complete."
msg77 = "\nFiltering Clones outside CDS."
msg78 = "\n#Error#Error clones annotation.Execution stop."
msg79 = "\nClones annotation complete."
msg80 = "#Error#Error during blastp analysis. Exit\n"
msg81 = "\nBlastp analysis complete."
msg82 = "\n#Error#Error during translation in six dataframe.Execution stop"
msg83 = "\nTranslation complete."
msg84 = "\n#Error#Chromosome name not given. Execution stop."
msg85 = "\n#Error#No input file given. Intesection error. Execution stop."
msg86 = "\nOnly one input file found. No intersection will be performed."
msg87 = "\n#Error#Chromosome Name is missing.Execution stop."
msg88 = '\n#Error#Error during merging files.Execution stop.'
msg89 = '\n#Error#Error in pybedtools. Function bedtools getfasta failed.Execution stop.'
msg90 = '\n#Error#Error in pybedtools. Function bedtools getfasta failed.Execution stop.'
msg91 = '\nPybedtools getfasta complete.'
msg92 = '\n#Error#Error while cleaning balstn output.Execution stop.'
msg93 = '\nBlastn output hash clean complete.'
msg94 = '#Error#Error conversion FastQ ==>FastA.Execution stop.\n'
msg95 = 'Conversion FastQ ==>FastA complete.\n'
msg96 = "#Error#Error conversion FastA ==>Tabular.Execution stop.\n"
msg97 = "Conversion FastA ==>Tabular complete.\n"
msg98 = "#Error#Error renaming Fasta Id.Execution stop.\n"
msg99 = "Renaming Fasta Id complete.\n"
msg100 = "#Error#Error conversion Tabular ==>FastA.Execution stop.\n"
msg101 = "\nConversion Tabular ==>FastA complete.\n"
msg102 = "\n#Error# Blastn Output filtering failed.Execution stop."
msg103 = "\nBlastn Output filtering complete."
msg104 = "\n#Error#Output Bed parsed failed.Execution stop."
msg105 = "\nOutput Bed parsed complete."
msg106 = "\n#Error#Error during clone cluster count.Execution stop."
msg107 = "\nClone cluster count.Complete"
msg108 = "\n#Error#Filtering cluster failed.Execution stop."
msg109 = "\nFiltering cluster complete."
msg110 = "\n#Error# Add gene description failed.Execution stop."
msg111 = "\nAdd gene description complete."
msg112 = "\n#Error#Add nucleotide sequence failed.Execution stop."
msg113 = "\nAdd nucleotide sequence complete."
msg114 = "\n#Error#Renaming chromosomes of reference fasta file failed.Execution stop."
msg115 = "\nRenaming chromosomes of reference fasta file complete"
msg116 = "#Error# Only primer 5\' forward was giveng.Please insert also primer 3\'.Execution stop."
msg117 = "#Error# Only primer 3\' forward was giveng.Please insert also primer 5\'.Execution stop."